### DISCUSSION TITLE
How to use analyzer

### DISCUSSION THREAD

**Discussion by fllay** [srsran/srsRAN_Project#403] [srsRAN-matlab] [answered]
Hi 
I am trying to use analyzer to view resource block. I config gnb with this option 

```log:
  filename: /tmp/gnb.log                                          # Path of the log file.
  all_level:  info                                               # Logging level applied to all layers.
  phy_rx_symbols_filename: /tmp/iq.log
```
Then I use this command in matlab

```
prach = nrPRACHConfig;
prach.DuplexMode = 'TDD';
prach.ConfigurationIndex = 159;
prach.SubcarrierSpacing = 30;
prach.LRA = 139;
prach.SequenceIndex = 1;
prach.PreambleIndex = 48;
srsPRACHAnalyzer(prach, '~/src/iq.log', 1391191, 3336);
```

I saw some plot and It seems to be ok (i.e. timing alignment information)

However, I have no clue how to use other function such as  `srsPUCCHAnalyzer, srsResourceGridAnalyzer`. I tried `help` in Matlab.

```
>> help srsResourceGridAnalyzer
 srsResourceGridAnalyzer Plots the heat map of a resource grid.
    srsResourceGridAnalyzer(NRBS, RGFILENAME, RGOFFSET, RGSIZE) displays the
    content (amplitude) of a resource grid of NRBS resource blocks (frequency
    domain) and one slot (time domain). RGOFFSET and RGSIZE are the offset
    and size, as a number of single-precision complex floating point numbers,
    of the slot to be analyzed inside the binary file.
 
    RG = srsResourceGridAnalyzer(...) also returns a matrix RG with the complex-
    valued samples of the resource grid (rows are subcarriers, columns are symbols).

>> help srsPUCCHAnalyzer
 srsPUCCHAnalyzer Analyzes a PUCCH transmission from a resource grid.
    srsPUCCHAnalyzer(CARRIER, PUCCH, RGFILENAME, RGOFFSET, RGSIZE) analyzes the
    NR Physical Uplink Control Channel transmission detailed by the carrier
    configuration CARRIER and by the control channel configuration PUCCH (either
    Format 1 or Format 2). The resource grid IQ samples are stored in the
    binary file RGFILENAME. RGOFFSET and RGSIZE are the offset and size, as
    a number of single-precision complex floating point numbers, of the slot
    containing the PUCCH transmission inside the binary file.
 
    Example:
    % The relevant data can be found in the log file generated by the srsRAN gNB:
    % look for lines like the following
    %    2023-06-07T20:54:24.502414 [Upper PHY] [I] [  584.19] RX_SYMBOL: sector=0 offset=636504 size=45864
    %    2023-06-07T20:54:24.502277 [UL-PHY2 ] [D] [  584.19] PUCCH: rnti=0x4601 format=1 ...
 
    % Use srsParseLogs to populate the carrier and PUCCH configuration objects
    % (you will be asked to select the PUCCH entry of the logs):
    [carrier, pucch] = srsParseLogs
 
    % Launch the analyzer
    srsPUCCHAnalyzer(carrier, pucch, 'rx_symbols.bin', 636504, 45864)
 
    See also srsParseLogs.
```
It seems that `srsPUCCHAnalyzer` needs infomation from `srsParseLogs` and from `srsParseLogs` help

```
 srsParseLogs Parses a PHY uplink channel log entry.
    [CARRIER, PHYCH, EXTRA] = srsParseLogs asks the user to select a section of
    the srsGNB logs corresponding to PHY uplink channel, parses the information
    and returns the following objects:
    CARRIER - an nrCarrierConfig object with the carrier configuration
    PHYCH   - a PHY uplink channel configuration object (specifically, an
              nrPUSCHConfig, an nrPUCCH1Config or an nrPUCCH2Config object)
    EXTRA   - a struct with PUSCH additional information, namely redundancy
              version, transport block size and target code rate (for PUCCH, the
              struct is empty).
 
    As an example of a log entry expected by the srsParseLogs function, the
    following excerpt from a srsGNB log file refers to a PUSCH transmission
    (similar ones can be found for PUCCH transmissions, too).
    Important: The srsGNB log level must be set to "debug" to obtain detailed
    information as shown below.
```

It seems that `srsParseLogs` requires the user to select some infomation in the srsGNB logs. I have no idea which part of the srsGNB logs to be selected? All I got now is an IQ.log which is a binary file and a gnb.log file. Please give me more information on how to use these tools.

**Comment by dvdgrgrtt** [comment, marked as answer]
Hi @fllay, thanks for looking into the analyzers.

The `srsResourceGridAnalyzer` is very simple, you only need to pick an uplink slot from the log, any of the lines like the following one
```
2023-10-08T19:14:54.738749 [Upper PHY] [I] [  690.17] RX_SYMBOL: sector=0 offset=79705 size=8568
```
which show the offset and size you have to provide to the function (in a similar way to what you do for the PRACH).

As for the PUSCH and PUCCH analyzers, they're a bit more complicated. First of all, you need to set the PHY log level to debug, otherwise you won't get all the data you need. Then, you have to look for the PUSCH transmission you want to analyze and find the corresponding transmitting slot - for instance, the following PUSCH is associated to the slot of the previous example (see, they have the same slot number 690.17)
```
2023-10-08T19:14:54.738854 [UL-PHY1 ] [D] [  690.17] PUSCH: rnti=0x4601 h_id=0 prb=[3, 6) symb=[0, 14) mod=QPSK rv=0 tbs=11 crc=OK iter=1.0 sinr=20.1dB t=182.0us
  rnti=0x4601
  h_id=0
  ...
```
This is what you need to select when the `srsParseLogs` helper function requests you to do so, all the way down until you get a new log entry. For the PUCCH, the procedure is similar.

We are currently preparing a tutorial that will be available together with the rest of [the documentation](https://docs.srsran.com/projects/project/en/latest/). In the meantime, I'm attaching a [couple of slides](https://github.com/srsran/srsRAN_Project/files/13939992/analyzer.pdf) that may be helpful - let me know if you need more help!

**Reply by fllay** [reply]
Thank you very  much. After I follow the instruction in the slides, I got it working now.

**Reply by pgawlowicz** [reply]
@fllay You might be interested in checking our new [MATLAB Testing Tools](https://docs.srsran.com/projects/project/en/latest/tutorials/source/matlab/source/index.html) tutorial.
